### DISCUSSION TITLE
Serving TW5 in existing NodeJS application

### DISCUSSION THREAD

**Discussion by Keyholder** [TiddlyWiki/TiddlyWiki5#8006] [Q&A] [answered]
Hey, guys, noob here, so thank you for patience. What if I have an existing NodeJS application (single-user, for localhost access) that I want to serve an **updatable** TiddlyWiki instance as one of its pages - what parts of your api should I be looking at? To make it more interesting, what would be that most canonical way provide access to custom API of this existing application to the code of TiddlyWiki modules in such case? Speaking simply, I'm talking here about somehow injecting TW5 as a plugin to existing node-based platform with reasonable level of mutual integration - this should be possible, I believe.

**Comment by Jermolene** [comment]
Hi @Keyholder that is an interesting idea, and I don't think anyone has done any work in that area yet.

Doing it elegantly would probably require some modest core modifications: for example, one approach would be to expose TiddlyWiki's server request handler so that the containing app could delegate requests to it.

Are there any established mechanisms in this area? For example, do  Express apps have a way to plugin route handlers?

**Reply by Keyholder** [reply]
I'm honestly surprised to be the first one asking about this use case - it seems such wiki could be obviously useful plugin for pretty much anything you are running locally on node.

As for established mechanisms... From what I see on https://expressjs.com/en/guide/routing.html - you pretty much said it yourself, `express.Router` just wraps server request handlers.

It would be really cool to have TiddlyWiki not only as console app, but as a library too.

**Reply by Jermolene** [reply]
Hi @Keyholder there's a demo here of using TiddlyWiki 5 as a library in another app. It's straightforward to spin up a wiki and then render tiddlers to HTML:

https://github.com/Jermolene/TiddlyWiki5DemoApp

Going back to the OP, it's probably not worth trying to reuse TW5's route handlers – the most important one is only a couple of lines; it's the one that returns a fully built wiki:

https://github.com/Jermolene/TiddlyWiki5/blob/9079186b18f25f435c987957d931858595fd2ac4/core/modules/server/routes/get-index.js#L15-L25

**Reply by Keyholder** [reply]
Oh, that's the example I was looking for! Thanks, it seems even easier then I expected.

About handlers reuse - well, for my current use case I need editable instance and not just rendering of given tiddlers. So, as far as I see, this way I'd need to recreate several handlers from `server/routes` - not sure this would be easier then root handler reuse.

**Reply by Jermolene** [reply, marked as answer]
Hi @Keyholder if you want to experiment with re-using the route handlers, you can get a reference to them with the following code:

```js
$tw.modules.forEachModuleOfType("route", function(title,routeDefinition) {
	console.log(routeDefinition);
});
```

You'd need to fake a "state" object to pass to the route handlers along with the request and response objects.

**Reply by Keyholder** [reply]
Thank you! Love this architecture - everything is at reach!

**Comment by linonetwo** [comment]
> I don't think anyone has done any work in that area yet. 

@Jermolene I've done this! TidGi Desktop is now starting a nodejs wiki instance, and redirect its output to IPC channel. In @Keyholder 's case, you just need to redirect it to your existing nodejs service's HTTP route.

Firstly, create a headless wiki (without its own http route) and store in `this.wikiInstance`

https://github.com/tiddly-gittly/TidGi-Desktop/blob/4e3985040f1d0324b1df0d54a3d83fd6b4876ab9/src/services/wiki/wikiWorker/startNodeJSWiki.ts#L112-L156

```ts
const wikiInstance = TiddlyWiki();
// other setup, see links above, and remember don't use "plugins/tiddlywiki/tiddlyweb" here
wikiInstance.boot.startup({ bootPath: TIDDLYWIKI_PACKAGE_FOLDER });
```

Then use `$tw.wiki.renderTiddler` to get wiki from instance, and feed to your service

https://github.com/tiddly-gittly/TidGi-Desktop/blob/4e3985040f1d0324b1df0d54a3d83fd6b4876ab9/src/services/wiki/wikiWorker/ipcServerRoutes.ts#L92-L96

```ts
const wikiHTML = this.wikiInstance.wiki.renderTiddler('text/plain', rootTiddler);
return { statusCode: 200, headers: { 'Content-Type': 'text/html' }, data: wikiHTML };
```

**Reply by Keyholder** [reply]
Thank you very much! Practical example of embedding TW5 as external dependency in another node project - that's exactly what I needed!
